flasc/analysis/expected_power_analysis_utilities.py

```diff
@@ -0,0 +1,484 @@
+
+from flasc.logging_manager import LoggingManager
+from flasc.utilities.energy_ratio_utilities import add_wd_bin, add_ws_bin
+
+logger_manager = LoggingManager()  # Instantiate LoggingManager
+logger = logger_manager.logger  # Obtain the reusable logger
+
+
+def _add_wd_ws_bins(
+    df_: pl.DataFrame,
+    wd_cols: List[str],
+    ws_cols: List[str],
+    wd_step: float = 2.0,
+    wd_min: float = 0.0,
+    wd_max: float = 360.0,
+    ws_step: float = 1.0,
+    ws_min: float = 0.0,
+    ws_max: float = 50.0,
+) -> pl.DataFrame:
+    """Add wind direction (wd) and wind speed (ws) bin columns to the dataframe.
+
+    Args:
+        df_ (pl.DataFrame): A polars dataframe, exported from a_in.get_df()
+        wd_cols (List[str]): A list of column names for wind direction
+        ws_cols (List[str]): A list of column names for wind speed
+        wd_step (float): The step size for the wind direction bins. Defaults to 2.0.
+        wd_min (float): The minimum wind direction value. Defaults to 0.0.
+        wd_max (float): The maximum wind direction value. Defaults to 360.0.
+        ws_step (float): The step size for the wind speed bins. Defaults to 1.0.
+        ws_min (float): The minimum wind speed value. Defaults to 0.0.
+        ws_max (float): The maximum wind speed value. Defaults to 50.0.
+
+    Returns:
+        pl.DataFrame: A polars dataframe with the wd and ws bin columns added.
+    """
+    df_ = add_ws_bin(df_, ws_cols, ws_step, ws_min, ws_max)
+    df_ = add_wd_bin(df_, wd_cols, wd_step, wd_min, wd_max)
+    return df_
+
+
+def _bin_and_group_dataframe_expected_power(
+    df_: pl.DataFrame,
+    test_cols: List[str],
+    bin_cols_without_df_name: List[str] = ["wd_bin", "ws_bin"],
+) -> pl.DataFrame:
+    """Group dataframes by bin columns and calculate the mean and variance of the test columns.
+
+    Args:
+        df_ (pl.DataFrame): A polars dataframe, exported from a_in.get_df()
+        test_cols (List[str]): A list of column names to calculate the mean and variance of
+        bin_cols_without_df_name (List[str]): A list of column names to bin the dataframes by.
+            Defaults to ["wd_bin", "ws_bin"].
+
+    Returns:
+        pl.DataFrame: A polars dataframe with the mean and variance of the test columns
+            grouped by bin columns.
+    """
+    num_df = df_["df_name"].n_unique()
+
+    bin_cols_with_df_name = bin_cols_without_df_name + ["df_name"]
+
+    # Group df_
+    df_ = (
+        df_.filter(pl.all_horizontal(pl.col(bin_cols_with_df_name).is_not_null()))
+        # Drop rows where all test_cols are null
+        .filter(pl.any_horizontal([pl.col(c).is_not_null() for c in test_cols]))
+        # Select for all bin cols present
+        .group_by(bin_cols_with_df_name, maintain_order=True)
+        # # Compute the mean and standard deviation of each of test_cols
+        .agg(
+            [pl.mean(c).alias(f"{c}_mean") for c in test_cols]
+            + [pl.var(c).alias(f"{c}_var") for c in test_cols]
```

Comment:
We can remove variance here since we're only using variance form the separate covariance matrix now.

Reply:
Ok done!